compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -6406,7 +6406,10 @@ private void rewriteInvocation(BLangInvocation invocation, boolean async) {
         // Reorder the arguments to match the original function signature.
         reorderArguments(invocation);
 
+        boolean isExprAndRequiredArgSame = !invocation.requiredArgs.isEmpty() &&
+                invocation.requiredArgs.get(0) == invocation.expr;
         rewriteExprs(invocation.requiredArgs);
+        BLangExpression firstRequiredArg = isExprAndRequiredArgSame ? invocation.requiredArgs.get(0) : null;
```

Comment:
Still, I am a bit unclear about this approach. As I understood what we have to do is, if first arg and expr are the same, we need to desugar either first arg or expr. If the first arg desugared via QueryDesugar it becomes a new node.
So after the L:6411 can't we just assign `invocation.expr = invocation.requiredArgs.get(0)` if isExprAndRequiredArgSame is true and remove rest of the changes?

Please check whether this works.

Reply:
https://github.com/ballerina-platform/ballerina-lang/blob/35169c467dc78f97ec3a6633f3fe9ded0aecd86c/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L6422 here expression is used if we desugared before this step we can not use it here

https://github.com/ballerina-platform/ballerina-lang/blob/35169c467dc78f97ec3a6633f3fe9ded0aecd86c/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L8328

Reply:
Also, instead of checking whether expr == requieredArgs.get(0), check whether we have a specific flag of some kind. eg: for langlib invocations, we add the expr as a requiredArg. There might be a flag to filter this. Please check

Reply:
Updated e93c12f6b848f94a54b1ac2bdd219e3ade85cc94 , 4a3848fda70a488df79928476c2ce500b03d1c5f